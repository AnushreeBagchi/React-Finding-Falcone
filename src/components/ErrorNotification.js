import React from "react";
import {resetError} from "../store/actions/errors";
import { connect } from "react-redux";

class ErrorNotification extends React.Component{
    
    ontryAgainClick = () => {
        this.props.history.push({
          pathname: "/"
        });
        this.props.resetError();
      };

    render() {
        const error = this.props.location.state.response.error.error;
        return (
            <>
              {error && (
                <div className="error">
                  <span>{error}</span>
                  <button onClick={this.ontryAgainClick}>Try Again</button>
                </div>
              )}
            </>
          );
    };
}
const mapStateToProps = (state) => ({
    state,
  });
  const mapDispatchToProps = (dispatch) => ({
    resetError: () => {
      dispatch(resetError());
    }
  });
  
  export default connect(mapStateToProps, mapDispatchToProps)(ErrorNotification);
  