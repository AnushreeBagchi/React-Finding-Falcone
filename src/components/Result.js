import React from "react";
import "../css/AppStyle.css";
import PropTypes from "prop-types";
import {SUCCESS_MSG, FAILED_MSG} from "../store/constants"

class Result extends React.Component {
  static propTypes = {
    history: PropTypes.object,
  };

  ontryAgainClick = () => {
    this.props.history.push({
      pathname: "/"
    });
  };

  render() {
    const result = this.props.location.state.response.result;
    const isSuccess =
      result && result.status === "success"
        ? true
        : false;
    const successMsg = SUCCESS_MSG;
    const failureMsg = FAILED_MSG;
    const planet = result && result.planet_name;

    return (
      <div className="resultDiv">
        {!isSuccess ? (
          <h2>{failureMsg}</h2>
        ) : (
          <>
            <h2>{successMsg}</h2>
            <h3>Time Taken :{this.props.history.location.state.timetaken}</h3>
            <h3>Planet found: {planet}</h3>
          </>
        )}
        <button onClick={this.ontryAgainClick}>Start Again</button>
      </div>

    );
  }
}

export default Result;
