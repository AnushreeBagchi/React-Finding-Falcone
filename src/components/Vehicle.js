import React from "react";
import "../css/AppStyle.css";
class Vehicle extends React.Component {
  onVehicleSelect = (event) => {
    let distance = this.props.selectedPlanet[0].distance;
    let vehicleObj = this.props.vehicles.filter(veh => veh.name === event.target.value);
    let speed = vehicleObj[0].speed;
    let time = distance/speed;
    this.props.onVehicleSelect(event, time);
  }
  render() {
    const vehicles = this.props.vehicles;
    return (
      <div className="vehicleList">
        {vehicles.map((vehicle) => {
          let index = vehicles.indexOf(vehicle);
          let isRangeLess =
            this.props.selectedPlanet[0].distance > vehicle.max_distance;
          return (
            <div key={index}>
              <input
                key={vehicles.indexOf(vehicle)}
                type="radio"
                value={vehicle.name}
                name={this.props.destinationGroup}
                onClick={this.onVehicleSelect}
                speed = {vehicle.speed}
                disabled={vehicle.total_no === 0 || isRangeLess ? true : false}
              />
              {`${vehicle.name} (${vehicle.total_no})`}
            </div>
          );
        })}
      </div>
    );
  }
}
export default Vehicle;
