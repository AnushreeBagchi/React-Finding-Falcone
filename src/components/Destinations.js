import React from "react";
import "../css/AppStyle.css";
import Vehicle from "../components/Vehicle";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { addDestination } from "../store/destinations";

class Destinations extends React.Component {
  static propTypes = {
    planets: PropTypes.array,
    vehicles: PropTypes.array,
    destinationSelected: PropTypes.func,
    vehicleSelected: PropTypes.func,
    index: PropTypes.string,
    destinations: PropTypes.object,
    showVehicle: PropTypes.bool,
  };

  state = {
    selectedPlanet: {},
  };
  onDestinationChange = (event) => {
    let selectedPlanet = this.props.planets.filter(
      (planet) => planet.name === event.currentTarget.value
    );
    this.setState({ selectedPlanet });
    this.props.addDestination({destination : this.props.index, value: event.currentTarget.value})
  };

  render() {
    let planets = this.props.planets;
    let optionItems = planets.map((planet) => (
      <option key={planet.name}>{planet.name}</option>
    ));
    let selectedValue =
      this.props.destinations[this.props.index].selectedPlanet ||
      "Select Destination";

    return (
      <div className="destination">
        <h6> {this.props.index.toUpperCase()} </h6>
        <select onChange={this.onDestinationChange}>
          <option value={selectedValue}>{selectedValue}</option>
          {optionItems}
        </select>
        {this.props.showVehicle && (
          <Vehicle
            vehicles={this.props.vehicles}
            destinationGroup={this.props.index}
            selectedPlanet={this.state.selectedPlanet}
          ></Vehicle>
        )}
      </div>
    );
  }
}
// export default Destinations;

const mapStateToProps = state => ({
  state
});
const mapDispatchToProps = dispatch => ({
  addDestination : (data) => {dispatch(addDestination(data))}
})

export default connect(mapStateToProps, mapDispatchToProps)(Destinations);
