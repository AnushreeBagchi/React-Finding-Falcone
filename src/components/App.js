import React from "react";
import Destinations from "./Destinations";
import "../css/AppStyle.css";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import {fetchPlanets, getAvailablePlanets } from "../store/planets/planets";
import { fetchVehicles , getAvailableVehicles} from "../store/vehicles/vehicles";
import { getInitialDestinations } from "../store/destinations/destinations";
import { findFalcone } from "../store/findFalcone";


class App extends React.Component {
  static propTypes = {
    history: PropTypes.object,
  };
  
  async componentDidMount() {
    await this.props.fetchPlanets();
    await this.props.fetchVehicles();
    this.props.getInitialDestinations();
  }
  
  getTimeTaken = () => {
    let timetaken = 0;
    Object.keys(this.props.state.destinations).forEach((key) => {
      if (this.props.state.destinations[key]) {
        let curr_time = this.props.state.destinations[key].timetaken;
        if (curr_time) {
          timetaken += curr_time;
        }
      }
    });
    return timetaken;
  };

  onReset = () => {
    this.props.getInitialDestinations();
  };

  findFalcone = async () => {
    let selectedPlanets = [];
    let selectedVehicles = [];
    let destinations = this.props.state.destinations;
    Object.keys(destinations).forEach(key => {
      let dest = destinations[key];
      if (dest.selectedPlanet && dest.selectedVehicle ){
        selectedPlanets.push(dest.selectedPlanet);
        selectedVehicles.push(dest.selectedVehicle);
      }
    })

    let requestBody = {
      planet_names: selectedPlanets,
      vehicle_names: selectedVehicles,
    };
    this.props.findFalcone(requestBody).then(response => this.goToResult());
  };

  goToResult = () => {
    this.props.history.push({
      pathname: "/result/",
      state: {
        response: this.props.state.findFalcone,
        timetaken: this.getTimeTaken(),
      },
    });
  };

  render() {
    return (
      <div className="app">
        <h1 className="header">Finding Falcone!</h1>
        {this.props.state.planets.length > 0 && this.props.state.vehicles.length > 0 ? (
          <ul className="destinationList">
            {Object.keys(this.props.state.destinations).map((dest) => (
              <Destinations
                key={dest}
                index={dest}
                planets={getAvailablePlanets(this.props.state)}
                vehicles={getAvailableVehicles(this.props.state)}
              ></Destinations>
            ))}
          </ul>
        ) : (
          <p></p>
        )}
        <div>Time Taken : {this.getTimeTaken()}</div>
        <button className="resetButton" onClick={this.onReset}>
          Reset
        </button>
        <button className="searchButton" onClick={this.findFalcone}>
          Find Falcone!
        </button>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  state
});
const mapDispatchToProps = dispatch => ({
  fetchPlanets : () => { dispatch(fetchPlanets()) },
  fetchVehicles : () => {dispatch(fetchVehicles())},
  getAvailablePlanets : () => {dispatch(getAvailablePlanets())},
  getInitialDestinations : () => {dispatch(getInitialDestinations())},
  findFalcone : (data) => {return dispatch(findFalcone(data))}
})

export default connect(mapStateToProps, mapDispatchToProps)(App);