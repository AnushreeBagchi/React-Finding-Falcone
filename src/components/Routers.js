import React from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route
} from "react-router-dom";
import App from "./App";
import NotFound from "./NotFound";
import Result from "./Result";

export default function Routers() {
  return (
    <Router>
      <div>
        <Switch>
            <Route exact path="/" component={App}/>
            <Route exact path="/result" component={Result}/>
            <Route exact component={NotFound}/>
        </Switch>
      </div>
    </Router>
  );
}