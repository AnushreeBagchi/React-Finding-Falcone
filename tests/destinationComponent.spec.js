import { shallow, configure } from "enzyme";
import { Destinations } from "../src/components/Destinations";
import React from "react";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });
const getState = () => ({
    destinations: {
      destination1: {selectedPlanet: "Donlon",selectedVehicle : "Space pod"},
    },
    planets: [{ distance: 100, name: "Donlon" }],
    vehicles: [{ max_distance: 200, name: "Space pod", speed: 2, total_no: 2 }]
  });

describe("Testing Destinations component renders correctly", () => {
  it("Test", () => {
    const dest = "destination1";

    const wrapper = shallow(
      <Destinations
        index={dest}
        planets={getState().planets}
        vehicles={getState().vehicles}
        state={getState()}
      />
    );
    
    expect(wrapper).toMatchSnapshot();
  });
});
