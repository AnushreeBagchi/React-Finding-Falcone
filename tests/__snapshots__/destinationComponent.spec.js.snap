// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Destinations component Test 1`] = `ShallowWrapper {}`;
